examples/blockchain/src/blockchain.rs

```diff
@@ -0,0 +1,119 @@
+use codec::{Decode, Encode};
+use futures::{channel::mpsc::UnboundedReceiver, StreamExt};
+use log::info;
+
+#[derive(PartialEq, Eq, Clone, Debug, Decode, Encode, std::hash::Hash)]
+pub enum Transaction {
+    Print(u32, u32),
+    Burn(u32, u32),
+    Transfer(u32, u32, u32),
+}
+
+#[derive(Clone, Debug)]
+pub enum ProcessedTransaction {
+    Accepted(Transaction),
+    Rejected(Transaction),
+}
+
+#[derive(Clone, Debug)]
+pub struct State(Vec<u32>);
+
+impl State {
+    fn new(n_clients: u32) -> Self {
+        Self(vec![0; n_clients as usize])
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Block {
+    transactions: Vec<ProcessedTransaction>,
+    state: State,
+}
+
+#[derive(Debug)]
+pub struct Blockchain {
+    state: State,
+    transactions_buffer: Vec<Transaction>,
+    blocks: Vec<Block>,
+    n_blocks: u32,
+    n_tr_per_block: u32,
+    rx_transactions: UnboundedReceiver<Transaction>,
+}
+
+impl Blockchain {
+    pub fn new(
+        n_clients: u32,
+        n_blocks: u32,
+        n_tr_per_block: u32,
+        rx_transactions: UnboundedReceiver<Transaction>,
+    ) -> Self {
+        assert!(n_clients > 0 && n_blocks > 0 && n_tr_per_block > 0);
+        Self {
+            state: State::new(n_clients),
+            transactions_buffer: vec![],
+            blocks: vec![],
+            n_blocks,
+            n_tr_per_block,
+            rx_transactions,
+        }
+    }
+
+    fn add_transaction(&mut self, transaction: Transaction) -> bool {
+        self.transactions_buffer.push(transaction);
+        if self.transactions_buffer.len() == self.n_tr_per_block as usize {
+            self.create_block();
+        }
+        self.blocks.len() == self.n_blocks as usize
+    }
+
+    fn create_block(&mut self) {
+        let mut transactions: Vec<ProcessedTransaction> = vec![];
```

Comment:
`let transactions = self.transactions_buffer.drain(..).map(|tx| self.process_transaction(tx)).collect();`
This is more idiomatic and takes care of line 83.